tests/migrations/test_operations.py

```diff
@@ -5227,6 +5227,321 @@ def assertModelsAndTables(after_db):
+            actual_indexes = get_author_field_indexes()
+
+            # only MySQL implicitly optimizes FK indexes
+            fk_index_optimization_expected = connection.vendor == "mysql"
+            if fk_index_optimization_expected:
+                self.assertNotEqual(set(actual_indexes), set(initial_indexes))
+                # after the additional index found we need to clear field's indexes
+                # before the next checks
+                self.apply_operations(
+                    app_label,
+                    project_state,
+                    [
+                        migrations.RemoveField("Book", "author"),
+                        migrations.AddField("Book", "author", author_field),
+                    ],
+                )
+            else:
+                self.assertEqual(set(actual_indexes), set(initial_indexes))
+
+        assert_index_created(
+            [
+                migrations.AlterUniqueTogether("Book", ["author", "title"]),
+                migrations.AlterUniqueTogether("Book", []),
+            ]
+        )
+        assert_index_created(
+            [
+                migrations.AlterIndexTogether("Book", ["author", "title"]),
+                migrations.AlterIndexTogether("Book", []),
+            ]
+        )
+
+        unique_constraint = models.UniqueConstraint(
+            fields=["author", "title"],
+            name="author_title_uniq",
+        )
+        assert_index_created(
+            [
+                migrations.AddConstraint("Book", unique_constraint),
+                migrations.RemoveConstraint("Book", unique_constraint.name),
+            ]
+        )
+
+        simple_index = models.Index(
+            fields=["author", "title"],
+            name="author_title_index",
+        )
+        assert_index_created(
+            [
+                migrations.AddIndex("Book", simple_index),
+                migrations.RemoveIndex("Book", simple_index.name),
+            ]
+        )
+
+        if connection.features.supports_expression_indexes:
+            f_index = models.Index(
+                models.F("author"),
+                models.F("title"),
+                name="author_title_index",
+            )
+            assert_index_created(
+                [
+                    migrations.AddIndex("Book", f_index),
+                    migrations.RemoveIndex("Book", f_index.name),
+                ]
+            )
+
+        if not connection.features.supports_index_column_ordering:
+            negative_index = models.Index(
+                fields=["-author", "title"],
+                name="author_title_index",
+            )
+            assert_index_created(
+                [
+                    migrations.AddIndex("Book", negative_index),
+                    migrations.RemoveIndex("Book", negative_index.name),
+                ]
+            )
+
+    def test_composed_func_indexes_dont_create_implicit_fk_index_when_deleting(self):
```

Comment:
This tests works without a patch, we should add it in a separate commit/PR.

Reply:
makes sense. I've squashed everything related to the original bugfix in one commit and then supporting of the different kinds of indexes/constraints in another one. Does this work? Not sure that we should make another PR as in this case this PR will introduce non-working code